### ISSUE TITLE
Raw JSON Layer Design Proposal

### ISSUE THREAD

**Issue by klesh** [merico-dev/lake#1201]
# Raw JSON Layer design

## Why

Currently, on Tool Layer, for each plugin, we read data from API, extract and transform fields we need and store into predefined tables, this approach poses some limitations:

1. data lost: those fields that we didn't extract got throwed away.
2. data distortion: some fields got transformed to type or format that not exactly ideal.
3. development/test: in order to develop/test our collection logic, we have to manually run code against a real endpoint, and the responses from api endpoint varies from time to time.

which leads to following difficulties:

1. User: whenever a user set a wrong parameter for a collection task, he/she have to recollect data from api all over again, which is time consuming.
2. Test: no stable fake data for automated test, and QA verification. it's hard to setup and maintain a mock up deployment for each and every plugin.
3. Development: it's unnecessary to actually fetch data from api when we are just developing some simple transformation logic.

## How

In order to overcome these drawbacks, most of us had reached some agreements:

1. A `raw json layer` must be added before transformation to preserve the full information from api responses.
2. This layer should be managed by framework, plugin can saves/loads their `raw json` by using `functions` or `interfaces` provided by framework, plugin **should not** access `raw json tables` directly.
3. Each tool layer entity has its own respected `raw json table`, the relationship between them is `OneToOne`.

## What

I propose the following solution:

![image](https://user-images.githubusercontent.com/61080/153619771-ebc0015a-aadb-44a5-979c-2bfb72f61a53.png)

1. pseudocode for `jira issue collector`

   ```go
   func CollectJiraIssues(apiClient *JiraApiClient, source *JiraSource, boardId int) error {
       rawJiraIssueTable := RawDataTable(&RawJiraIssue{})
       rawJiraBoardIssueTable := RawDataTable(&RawJiraBoardIssue{})
       
       return apiClient.FetchPages(func(res *http.Response) error {
           if res.StatusCode != 200 {
               return fmt.Errorf("unexecpted status code")
           }
           jiraApiIssuesResponse := &JiraApiIssuesResponse{}
           err := core.UnmarshalResponse(res, jiraApiIssuesResponse)
           if err != nil {
               return err
           }
           // process issues
           for _, jiraApiIssue := range jiraApiIssuesResponse.Issues {
               err = rawJiraIssueTable.Store(&RawJiraIssue{
                   SourceId: source.ID,
                   IssueId: jiraApiIssue.Id,
                   RawData: jiraApiIssue,
               })
               if err != nil {
                   return err
               }
               rawJiraBoardIssueTable.Store(&RawJiraBoardIssue{
                   SourceId: source.ID,
                   BoardId: boardId,
                   IssueId: jiraApiIssue.Id,
               })
               if err != nil {
                   return err
               }
           }
       })
   }
   ```

   which store data into `raw_jira_issues` table like following:

   | source_id | issue_id   | raw_json                                         | created_at                    | updated_at                    |
   | --------- | ---------- | ------------------------------------------------ | ----------------------------- | ----------------------------- |
   | 1         | 1067629368 | {"id": 1067629368, "self": "...", "fields": ...} | 2022-01-26 09:52:17.697000000 | 2022-01-26 09:52:17.697000000 |
   | 1         | 1009381716 | {"id": 1009381716, "self": "...", "fields": ...} | 2022-01-26 09:52:17.697000000 | 2022-01-26 09:52:17.697000000 |

2. pseudocode for `jira_issue_transformer`

   ```go
   func TransformJiraIssue(source *JiraSource, boardId int) error {
       rawJiraIssueTable := RawDataTable(&RawJiraIssue{})
       return rawJiraIssueTable.Iterate(func(rawJiraIssue *RawJiraIssue) error {
           // extract information from rawJiraIssue and create/update JiraIssue accordingly
           return nil
       })
   }
   ```

   

By this approach, we can solve mentioned difficulties:

1. Users no long have to re-collect data from api if they set a wrong `customfield`
2. We can collect and copy those `raw json tables` for testing and QA verification, whenever a new type of data source came in, we just collect and added it to our copying list, no need to manually fake any data.
3. Speed up our new fields collection and transformation logic debug.

**Comment by mindlesscloud** [comment]
I have some suggestions about abstracting the process from plugins. By defining  3 interfaces `Paginator`, `RawJsonExtractor` and `Store`, we could fouce on the implementation of these interfaces.
```go
type Paginator interface {
	HasNext() bool
	GetNext() (*http.Response, error)
}

type RawJson struct {
	Id        string
	Blob      []byte
	CreatedAt time.Time
}
type RawJsonExtractor interface {
	Extract(resp *http.Response) ([]*RawJson, error)
}
type Store interface {
	Save(context.Context, []*RawJson) error
}

func Collect(ctx context.Context, paginator Paginator, extractor RawJsonExtractor, store Store) error{
	for paginator.HasNext() {
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
		}
		resp, err := paginator.GetNext()
		if err != nil {
			return err
		}
		objs, err := extractor.Extract(resp)
		if err != nil {
			return err
		}
		err = store.Save(ctx, objs)
		if err != nil {
			return err
		}
	}
	return nil
}
```

**Comment by klesh** [comment]
@mindlesscloud good idea, how to address `board_issues` relationship with this approach? this relationship can not be found anywhere inside the raw response, but merely from parameter that user passed in.

**Comment by warren830** [comment]
In my opinion, we can keep only sub task flow in framework, and put the interfaces(or say flow of specific sub task) of one sub task in plugins layer. We still have a lot of data sources to integrate, and in large probability, there are lots of differences in realizing a sub task among multiple datasources. @mindlesscloud 's proposal can restrain the code structure by both inside and external developers, so it will be easier to maintain by us in the future. And in framework, we can pay more attention on how to manage sub tasks.

**Comment by warren830** [comment]
BTW, I still think it may bring too many tables if we use one table to map a kind of entity of one datasource.  For current datasources, in tool layer, if we store raw json for notes, comments, events separately, it will bring about 40 more tables, almost one datasource need 20 tables in total. With 50 more datasources, we may need hundreds of tables. In general, I think we'd better keep raw_json layer as simple as possible, and do more work when store the data.

**Comment by klesh** [comment]
> In my opinion, we can keep only sub task flow in framework, and put the interfaces(or say flow of specific sub task) of one sub task in plugins layer. We still have a lot of data sources to integrate, and in large probability, there are lots of differences in realizing a sub task among multiple datasources. @mindlesscloud 's proposal can restrain the code structure by both inside and external developers, so it will be easier to maintain by us in the future. And in framework, we can pay more attention on how to manage sub tasks.

This can be true if and only if the `interface` is **complete**, so it is stable and never need to be changed. which is quite like the concept of `zero friction` in physis, no so practical in reality. So, Whenever we change the interface a little bit, it breaks existing 3rd plugins.

On the other hand, the `library` approach has no such issue, because code got compiled into plugin and run on its own, plugin developer can decide when to upgrade their code, and users can still use existing outdated plugins as they see fit.

As always, I stand against using `interface` as a norm, check the git history of https://github.com/merico-dev/lake/blame/main/plugins/core/plugin.go this simple interface has been modifed couple of times. This would not be ok if we had some impactful 3rd plugins in market.

**Comment by klesh** [comment]
> BTW, I still think it may bring too many tables if we use one table to map a kind of entity of one datasource. For current datasources, in tool layer, if we store raw json for notes, comments, events separately, it will bring about 40 more tables, almost one datasource need 20 tables in total. With 50 more datasources, we may need hundreds of tables. In general, I think we'd better keep raw_json layer as simple as possible, and do more work when store the data.

Can you post the detail of your solution to this matter? What do you suggest we do in order to keep it simple?

**Comment by klesh** [comment]
# New Design  2022/2/25

## Summary of opinions

- Be more general
- Single responsibility principle
- Framework in control, declarative plugin task approach
- Future oriented, consider tasks management, multi api version compatibility


```go
func CollectApiIssues(
	apiClient *JiraApiClient,
	source *models.JiraSource,
	boardId uint64,
	since time.Time,
) error {
	// --------- skip this part begin --------
	// user didn't specify a time range to sync, try load from database
	if since.IsZero() {
		var latestUpdated models.JiraIssue
		err := lakeModels.Db.Where("source_id = ?", source.ID).Order("updated DESC").Limit(1).Find(&latestUpdated).Error
		if err != nil {
			logger.Error("jira collect issues:  get last sync time failed", err)
			return err
		}
		since = latestUpdated.Updated
	}
	// build jql
	jql := "ORDER BY updated ASC"
	if !since.IsZero() {
		// prepend a time range criteria if `since` was specified, either by user or from database
		jql = fmt.Sprintf("updated >= '%v' %v", since.Format("2006/01/02 15:04"), jql)
	}
	query := &url.Values{}
	query.Set("jql", jql)
	// --------- skip this part end --------

	const SIZE = 100

	collector, err := core.NewApiCollector(core.ApiCollectorArgs{
		ApiClient: apiClient,
		PageSize:  SIZE,
		/*
			url may use arbitrary variables from different source in any order, we need GoTemplate to allow more
			flexible for all kinds of possibility.
			Pager contains information for a particular page, calculated by ApiCollector, and will be passed into
			GoTemplate to generate a url for that page.
			We want to do page-fetching in ApiCollector, because the logic are highly similar, by doing so, we can
			avoid duplicate logic for every tasks, and when we have a better idea like improving performance, we can
			do it in one place
		*/
		UrlTemplate: "agile/1.0/board/{{ Params.BoardId }}/issue?startAt={{ Pager.StartAt }}&maxResults={{ Pager.Size }}",
		/*
			Some api might do pagination by http headers
		*/
		//Header: func(collector *core.ApiCollector) http.Header {
		//},
		/*
			Sometimes, we need to collect data based on previous collected data, like jira changelog, it requires
			issue_id as part of the url.
			We can mimic `stdin` design, to accept a `Input` function which produces a `Iterator`, collector
			should iterate all records, and do data-fetching for each on, either in parallel or sequential order
			UrlTemplate: "api/3/issue/{{ Input.ID }}/changelog"
		*/
		//Input: databaseIssuesIterator,
		/*
			This struct will be JSONEncoded and stored into database along with raw data itself, to identity minimal
			set of data to be process, for example, we process JiraIssues by Board
		*/
		Params: JiraApiParams{
			SourceId: source.ID,
			BoardId:  boardId,
		},
		/*
			Reduce records come out from api
		*/
		Query: query,
		/*
			ApiCollector will try to `UnmarshalResponse` into this variable, and handle failure
		*/
		ResponseData: &JiraApiIssuesResponse{},
		/*
			Accept response and Unmarshalled body, return raw data either a single object or list
		*/
		OnData: func(res *http.Response, body interface{}) (interface{}, error) {
			issuesBody := body.(*JiraApiIssuesResponse)
			return issuesBody.Issues, nil
		},
		/*
			For api endpoint that returns number of total pages, ApiCollector can collect pages in parallel with ease,
			or other techniques are required if this information was missing.
		*/
		GetTotalPages: func(res *http.Response, body interface{}) (int, error) {
			issuesBody := body.(*JiraApiIssuesResponse)
			return issuesBody.Total / SIZE, nil
		},
		/*
			Table store raw data
		*/
		Table: "jira_api_issues",
	})

	if err != nil {
		return err
	}

	return collector.Execute()
}
```

@hezyin

**Comment by hezyin** [comment]
@klesh This is definitely going in the right direction. Let me quickly summarize my understanding and questions:

### Understanding of current design

- `ApiClient` used to be responsible for handling pagination and parallelizing requests. It seems in the new design both responsibilities are extracted out of `ApiClient` and assigned to `ApiCollector`.
- How this works is that we tell `ApiCollector` what API to query, what parameters to use, how to handle pagination (if any) and `ApiCollector` will take care of the collection process and store API responses in a database.

### Questions for current design

1. Does `Query` get appended to `UrlTemplate`? If so, why keep them separate?
2. Why do we unmarshal API response into a struct? Is this for data validation purposes? I'm thinking that `ResponseData` and `OnData` may belong somewhere else or at least we need some more explanation/justification.
3. For pagination, what happens if API doesn't return total pages, do we fall back to get next page strategy or something else?

**Comment by klesh** [comment]
> @klesh This is definitely going in the right direction. Let me quickly summarize my understanding and questions:
> 
> ### Understanding of current design
> * `ApiClient` used to be responsible for handling pagination and parallelizing requests. It seems in the new design both responsibilities are extracted out of `ApiClient` and assigned to `ApiCollector`.
> * How this works is that we tell `ApiCollector` what API to query, what parameters to use, how to handle pagination (if any) and `ApiCollector` will take care of the collection process and store API responses in a database.
> 
> ### Questions for current design
> 1. Does `Query` get appended to `UrlTemplate`? If so, why keep them separate?
> 2. Why do we unmarshal API response into a struct? Is this for data validation purposes? I'm thinking that `ResponseData` and `OnData` may belong somewhere else or at least we need some more explanation/justification.
> 3. For pagination, what happens if API doesn't return total pages, do we fall back to get next page strategy or something else?

Yes, you got it right.

Answers to Questions:
1. Yes, it does. `Query` is separated because it is optional, it is easier for plugin task to pass down query string if any, we can then do staff like omitting the `query key` altogether if value was missing.
2. Validation and error handling, `unmarshal` logic is needed for pretty much every collection task, by doing so, we can do it in one place to avoid error-swallowing, and provide a unity error interace for users.
3. Yes, maybe page by page or step concurrency technique, haven't quite decided yet, but we can keep improve the strategy in `ApiCollector` without touching plugin tasks.

**Comment by hezyin** [comment]
@klesh 

1. Whether we separate query string with URL or not is a minor design choice so I don't have a strong preference. I'm leaning towards keeping them together since most web crawling frameworks I've used (e.g., scrapy in python) only ask for a single URL and it's the user's responsibility to build that URL.
2. If we agree on the following responsibility definition for `ApiCollector`, then I think we should delay the `unmarshal` logic to a later point, possibly when developers process the stored API response. `ApiCollector` is only concerned with where to send API requests (URL), how to handle pagination and parallelism, and checking that we're getting HTTP 200 for those API requests. `ApiCollector` should be oblivious to the internal data structure.

> How this works is that we tell ApiCollector what API to query, what parameters to use, how to handle pagination (if any) and ApiCollector will take care of the collection process and store API responses in a database.

**Comment by klesh** [comment]
Ok, I concur.

**Comment by klesh** [comment]
Updated design:

1. `Query` function is optional, developer can plug them into `UrlTemplate` directly, or use this function when url is getting too long to maintain
2. `ApiCollector` is not doing `unmarshal` now

```go
func CollectApiIssues(
	apiClient *JiraApiClient,
	source *models.JiraSource,
	boardId uint64,
	since time.Time,
) error {
	// --------- skip this part begin --------
	// user didn't specify a time range to sync, try load from database
	if since.IsZero() {
		var latestUpdated models.JiraIssue
		err := lakeModels.Db.Where("source_id = ?", source.ID).Order("updated DESC").Limit(1).Find(&latestUpdated).Error
		if err != nil {
			logger.Error("jira collect issues:  get last sync time failed", err)
			return err
		}
		since = latestUpdated.Updated
	}
	// build jql
	jql := "ORDER BY updated ASC"
	if !since.IsZero() {
		// prepend a time range criteria if `since` was specified, either by user or from database
		jql = fmt.Sprintf("updated >= '%v' %v", since.Format("2006/01/02 15:04"), jql)
	}
	// --------- skip this part end --------

	const SIZE = 100

	collector, err := core.NewApiCollector(core.ApiCollectorArgs{
		ApiClient: apiClient,
		PageSize:  SIZE,
		/*
			url may use arbitrary variables from different source in any order, we need GoTemplate to allow more
			flexible for all kinds of possibility.
			Pager contains information for a particular page, calculated by ApiCollector, and will be passed into
			GoTemplate to generate a url for that page.
			We want to do page-fetching in ApiCollector, because the logic are highly similar, by doing so, we can
			avoid duplicate logic for every tasks, and when we have a better idea like improving performance, we can
			do it in one place
		*/
		UrlTemplate: "agile/1.0/board/{{ Params.BoardId }}/issue",
		/*
			(Optional) Return query string for request, or you can plug them into UrlTemplate directly
		*/
		Query: func(pager *core.Pager) (*url.Values, error) {
			query := &url.Values{}
			query.Set("jql", jql)
			query.Set("startAt", fmt.Sprintf("%v", pager.Skip))
			query.Set("maxResults", fmt.Sprintf("%v", pager.Size))
			return query, nil
		},
		/*
			Some api might do pagination by http headers
		*/
		//Header: func(pager *core.Pager) http.Header {
		//},
		/*
			Sometimes, we need to collect data based on previous collected data, like jira changelog, it requires
			issue_id as part of the url.
			We can mimic `stdin` design, to accept a `Input` function which produces a `Iterator`, collector
			should iterate all records, and do data-fetching for each on, either in parallel or sequential order
			UrlTemplate: "api/3/issue/{{ Input.ID }}/changelog"
		*/
		//Input: databaseIssuesIterator,
		/*
			This struct will be JSONEncoded and stored into database along with raw data itself, to identity minimal
			set of data to be process, for example, we process JiraIssues by Board
		*/
		Params: JiraApiParams{
			SourceId: source.ID,
			BoardId:  boardId,
		},
		/*
			Accept response, return raw data either a single object or list
		*/
		OnData: func(res *http.Response) (interface{}, error) {
			issuesBody := &JiraApiIssuesResponse{}
			err := core.UnmarshalResponse(res, issuesBody)
			if err != nil {
				return nil, err
			}
			return issuesBody.Issues, nil
		},
		/*
			For api endpoint that returns number of total pages, ApiCollector can collect pages in parallel with ease,
			or other techniques are required if this information was missing.
		*/
		GetTotalPages: func(res *http.Response) (int, error) {
			issuesBody := &JiraApiIssuesResponse{}
			err := core.UnmarshalResponse(res, issuesBody)
			if err != nil {
				return 0, err
			}
			return issuesBody.Total / SIZE, nil
		},
		/*
			Table store raw data
		*/
		Table: "jira_api_issues",
	})

	if err != nil {
		return err
	}

	return collector.Execute()
}
```

**Comment by klesh** [comment]
Extractor Example

```go
func ExtractApiIssues(
	source *models.JiraSource,
	boardId uint64,
	since time.Time,
) error {
	extractor, err := core.NewApiExtractor(core.ApiExtractorArgs{
		Table:       "jira_api_issues",
		Params: JiraApiParams{
			SourceId: source.ID,
			BoardId:  boardId,
		},
		RowData: &JiraApiIssue{},
		Extractors: []core.RawDataExtractor{
			func(row interface{}, params interface{}) (interface{}, error) {
				p := params.(*JiraApiParams)
				issue := row.(*JiraApiIssue)
				issueId, err := strconv.ParseUint(issue.Id, 10, 64)
				if err != nil {
					return nil, err
				}
				return &models.JiraIssue{
					SourceId: p.SourceId,
					IssueId:  issueId,
					// other fields
				}, nil
			},
			func(row interface{}, params interface{}) (interface{}, error) {
				p := params.(*JiraApiParams)
				issue := row.(*JiraApiIssue)
				issueId, err := strconv.ParseUint(issue.Id, 10, 64)
				if err != nil {
					return nil, err
				}
				return &models.JiraBoardIssue{
					SourceId: p.SourceId,
					BoardId:  p.BoardId,
					IssueId:  issueId,
				}, nil
			},
		},
	})

	if err != nil {
		return err
	}

	return extractor.Execute()
}
```

**Comment by klesh** [comment]
Hi, @hezyin , I ran into some difficulty when trying to implement fetch pages in parallel
Check the `THIS WONT WORK` part, `res.Body` is a stream which can only be read once.
```go
func (collector *ApiCollector) fetchPagesAsync() error {
	// make sure table is created
	db := collector.args.Ctx.GetDb()
	err := db.Exec(GetRawTableCreationSql(collector.table)).Error
	if err != nil {
		return err
	}
	// flush old data
	err = db.Exec(GetRawTableDeletionSql(collector.table), collector.params).Error
	if err != nil {
		return err
	}

	if collector.args.GetTotalPages != nil {
		/* when total pages is available from api*/
		// fetch the very first page
		err := collector.fetchAsync(nil, func(res *http.Response) error {
			// gather total pages
			totalPages, err := collector.args.GetTotalPages(res)
			if err != nil {
				return err
			}
			// handle response body of first page
			err = collector.handleResponse(res) // THIS WONT WORK: because res.Body already got read in GetTotalPages
			if err != nil {
				return err
			}
			// fetch other pages in parallel
			for page := 2; page <= totalPages; page++ {
				err = collector.fetchAsync(&Pager{
					Page: page,
					Size: collector.args.PageSize,
					Skip: collector.args.PageSize * (page - 1),
				}, collector.handleResponse)
				if err != nil {
					return err
				}
			}
			return nil
		})
		if err != nil {
			return err
		}

	} else {
		/* when total pages is available from api*/
		// fetch page by page in sequential?
		// use step currency technique? fetch like 10 pages at once, if all went well, fetch next 10 pages?
		panic("not implmented")
	}
	return nil
}
```

**Comment by hezyin** [comment]
@klesh 

1. In order to read twice from `resp.body`, we can construct a new `io.ReadCloser` after reading once and hand that to the next reader, see the following example from [here](https://stackoverflow.com/questions/46948050/how-to-read-request-body-twice-in-golang-middleware):

```
bodyBytes, _ := ioutil.ReadAll(req.Body)
req.Body.Close()  //  must close
req.Body = ioutil.NopCloser(bytes.NewBuffer(bodyBytes))
```

2. `resp.body` needs to be closed after reading to avoid any resource leakage. Please check that we're properly closing them now.
3. It seems we only removed `OnResponse` from the `ApiCollector` interface but `OnData` is still there. Shall we also move `OnData` to somewhere else?
4. Now that we have made some progress with these examples, I suggest that we spend some time writing the documentation for the raw data layer, clearly explaining the interface and responsibilities of `ApiCollector` and `ApiExtractor`, and including an example usage. Note that these don't include implementation description, they're external-facing docs for developers wishing to implement their own `ApiCollector` and `ApiExtractor`.

**Comment by klesh** [comment]
@hezyin 
1. I see, we can make do with that. But, still, it will get `unmarshal` twice, do we really need to do it this way?
2. `core.UnmarshalResponse` closes it properly
3. I dont understand
4. Will do

**Comment by klesh** [comment]
closed by #1287
